web/packages/shared/components/AccessRequests/NewRequest/kube.ts

```diff
@@ -0,0 +1,94 @@
+/**
+ * Teleport
+ * Copyright (C) 2024 Gravitational, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+import { Attempt } from 'shared/hooks/useAttemptNext';
+
+import { PendingListItem } from './RequestCheckout';
+
+export type KubeNamespaceRequest = {
+  kubeCluster: string;
+  search: string;
+};
+
+/**
+ * Returns true if the currItem is a kube cluster or is a namespace
+ * of the currItem.
+ *
+ * Technically you can request for both a kube_cluster and its subresources
+ * but it's probably not what a user would expect from the web UI because
+ * requesting for subresources means narrowing access versus requesting
+ * access to the whole kube_cluster.
+ */
+export function excludeKubeClusterWithNamespaces(
+  currItem: PendingListItem,
+  allItems: PendingListItem[]
+) {
+  return (
+    currItem.kind !== 'kube_cluster' ||
+    !(
+      currItem.kind === 'kube_cluster' &&
```

Comment:
This check for `kind` is not necessary because we get to this point only if `currItem.kind !== 'kube_cluster` returns `false`.

But I see that this function is used to be meant when filtering a list of `PendingListItem`. In that case, it might be beneficial to make this function answer a "positive" question and then use it with a negation when filtering:

```typescript
export const isKubeClusterWithNamespaces = (
  item: PendingListItem,
  allItems: PendingListItem[]
) =>
  item.kind === 'kube_cluster' &&
  allItems.find(a => a.kind === 'namespace' && a.id == item.id);

// …

data
  .filter(d => d.kind !== 'role')
  .filter(d => !isKubeClusterWithNamespaces(d, data));
```

This will make it easier to compose with other functions in the future _and_ it can be used in more contexts than just filtering.

I don't think there's a convention for what a function starting with `exclude` should return, but it's rather well-established that a function that starts with `is` returns a boolean and answers a question regarding the value passed to it. For me, it's somewhat easier to understand than the version with `exclude`.